src/integrationTest/java/org/opensearch/security/rest/CompressionTests.java

```diff
@@ -69,72 +64,54 @@ public void testAuthenticatedGzippedRequests() throws Exception {
         try (final TestRestClient client = cluster.getRestClient(new BasicHeader("Content-Encoding", "gzip"))) {
-
-            final ForkJoinPool forkJoinPool = new ForkJoinPool(parallelism);
-
-            final Header basicAuthHeader = getBasicAuthHeader(ADMIN_USER.getName(), ADMIN_USER.getPassword());
-
-            final List<CompletableFuture<HttpResponse>> waitingOn = IntStream.rangeClosed(1, totalNumberOfRequests)
-                .boxed()
-                .map(i -> CompletableFuture.supplyAsync(() -> {
-                    final HttpPost post = new HttpPost(client.getHttpServerUri() + requestPath);
-                    post.setEntity(new ByteArrayEntity(compressedRequestBody, ContentType.APPLICATION_JSON));
-                    return i % 2 == 0 ? client.executeRequest(post) : client.executeRequest(post, basicAuthHeader);
-                }, forkJoinPool))
-                .collect(Collectors.toList());
-
-            final CompletableFuture<Void> allOfThem = CompletableFuture.allOf(waitingOn.toArray(new CompletableFuture[0]));
-
-            allOfThem.get(30, TimeUnit.SECONDS);
-
-            waitingOn.stream().forEach(future -> {
-                try {
-                    final HttpResponse response = future.get();
-                    assertThat(response.getBody(), not(containsString("json_parse_exception")));
-                    assertThat(response.getStatusCode(), anyOf(equalTo(HttpStatus.SC_UNAUTHORIZED), equalTo(HttpStatus.SC_OK)));
-                } catch (final Exception ex) {
-                    throw new RuntimeException(ex);
-                }
+            final CountDownLatch countDownLatch = new CountDownLatch(1);
+
+            final var authorizedRequests = AsyncActions.generate(() -> {
+                countDownLatch.await();
+                System.err.println("Generation triggerd authorizedRequests");
+                final HttpPost post = new HttpPost(client.getHttpServerUri() + requestPath);
+                post.setEntity(new ByteArrayEntity(compressedRequestBody, ContentType.APPLICATION_JSON));
+                return client.executeRequest(post, getBasicAuthHeader(ADMIN_USER.getName(), ADMIN_USER.getPassword()));
+            }, parallelism, totalNumberOfRequests);
+
+            final var unauthorizedRequests = AsyncActions.generate(() -> {
+                countDownLatch.await();
+                System.err.println("Generation triggerd unauthorizedRequests");
+                final HttpPost post = new HttpPost(client.getHttpServerUri() + requestPath);
+                post.setEntity(new ByteArrayEntity(compressedRequestBody, ContentType.APPLICATION_JSON));
+                return client.executeRequest(post);
+            }, parallelism, totalNumberOfRequests);
+
+            // Make sure all requests start at the same time
+            countDownLatch.countDown();
+
+            AsyncActions.getAll(authorizedRequests, 30, TimeUnit.SECONDS).forEach((response) -> {
```

Comment:
Will this interleave requests?

Reply:
Yup  - all requests are queued, waiting to start, they are all started at line 106, then individual blocks are waited for in the `getAll` lines - but they are still interweaved

Note, the countDownLatch might be overkill, but this way I'm sure all requests both success & failures are ready any waiting before any of them start.